Angular App

nickwxo opened the pull request:
This pull request will include the setting up of Angular routes for the various pages of the HarvestAPI website and well as the registration process.

Commit:
Merge pull request #5 from slashroots/api-core

API Core Integration

Commit:
Angular routing for website views for signup, signin, dashboard, user flow. Re-uses partials where possible and switches out content as necessary.

Commit:
The user sign up form is now able to create a user in the DB. So there's a new User controller and services.js to handle same. Also some fields for the User Schema to match up with Cam's Sign-up form. Form validation and password encryption not implemented yet - will do that today.

Commit:
Factory and select dropdown for the roles on signup form populating from DB.

Commit:
Forgot to implement the Factory for the roles so that the form dropdown could be populated and the selected role be assigned to the user upon submission. This commit implements the same. I still need to implement the form validation and password hashing.

Commit:
Sign-up form validation implemented. Passwords are encrypted using SHA-1 encryption.

Commit:
Visual cues for the form validation with red highlight and error messages for invalid fields.

Commit:
The user is now shown their chosen email address on the confirmation screen (instead of the lorem ipsum dummy text).

matjames007 committed:
Merge branch 'develop' into angular-ui-stuff

matjames007 commented on models/db.js:
```diff
@@ -39,6 +38,9 @@ var UserSchema = new Schema({
     us_contact: {type: String, required: true},
     us_user_creation_date: {type: Date, default: Date.now()},
     us_state: {type: String, default: 'pending'},
+    us_address: {type: String, default: 'pending'},
```
remove default attribute and replace with `required: true` for all mandatory fields.  The absence of the field will have a default of pending instead of throwing an error for missing field in request.

matjames007 commented on package.json:
```diff
@@ -10,13 +10,15 @@
     "passport-localapikey": "0.0.3",
-    "serve-favicon": "~2.2.0"
+    "serve-favicon": "~2.2.0",
+    "sha1": "^1.1.1"
```
This should be a part of the frontend application.  Hashing done on the server means that the password is sent over the wire in cleartext which is a security risk.

matjames007 commented on public/javascript/controllers/user-ctrl.js:
```diff
@@ -0,0 +1,42 @@
+ * Created by Nick on 15/09/16.
+ */
+
+angular.module('harvestv2')
```
Please comment.

matjames007 commented on public/javascript/controllers/user-ctrl.js:
```diff
@@ -0,0 +1,42 @@
+
+        $scope.createUser = function () {
+            if (!$scope.signupForm.$invalid && $scope.validatePasswordMatch($scope.user.us_password, $scope.passwordConfirm) && $scope.agreeChecked)
+                UserFactory.create($scope.user, function(user) {
```
Should be hashing the password here using SHA1.

matjames007 commented on public/javascript/controllers/user-ctrl.js:
```diff
@@ -0,0 +1,42 @@
+
+        $scope.setupVisualValidationCues = function () {
+
+            jQuery('#signupForm').validator();
```
Why use jQuery here?  Please remember we are writing an angular application.

matjames007 commented on public/javascript/services.js:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Created by nickjwill on 15/09/16.
+ */
+
```
Please comment.

matjames007 commented on routes/user/user.js:
```diff
@@ -87,6 +87,7 @@ exports.getUser = function(req, res, next) {
 exports.createUser = function(req, res, next) {
     var user = new User(req.body);
     user.us_activation_token = common.getRandomToken();
+    user.us_password = sha1(user.us_password);
```
Incorrect place to be hashing.  Please remove and do hashing on frontend application

matjames007 commented on public/partials/signup.html:
```diff
@@ -0,0 +1,101 @@
+                        </div>
+                        <div class="form-group">
+                            <label for="input8">Role Requested</label>
+                            <select ng-init="selectedRole=roles[0]" data-error="Please select a role!" class="form-control" ng-model="selectedRole" ng-change="user.us_user_role = selectedRole" ng-options="r._id as r.ro_role_name for r in roles" required>
```
Hey @nick22891 I believe the flow was for the user to be first assigned a default role - not to request it at the beginning.  Remove this field from the registration screen.  I will provide a path for all the application defaults.  This was missing from my endpoints.  

I think in the application defaults there will be information you can use such as:
`{app_title: "Harvest v2", app_author: "SlashRoots Foundation", default_role: "57d438497a48b2cb165fc9f0",...}`

matjames007 committed:
I have authentication into the application working - however this is part 1 of authentication.
Token based authentication is still required for the resources we are serving.
-Matthew Budram

matjames007 committed:
Manual Merge branch 'develop' into app-auth-impl

matjames007 committed:
Merge branch 'develop' into angular-ui-stuff

Also setup the endpoints with registration and emails.  This is now functional.

matjames007 committed:
Merge branch 'app-auth-impl' into angular-ui-stuff

matjames007 committed:
I have added in the ability to authenticate on the platform.

matjames007 commented:
The application can now register a user and login.  There are many dead links in the application and these require attention.
